Adding Header authentication to Gateway  (Apache)

marwanehcine opened the pull request:
On this PR, we add a new filter which allows to authenticate users using the HTTP headers received from Apache. A new LDAP account is then created if `isCreateNonExistingUsersInLDAP` property is `true`.

This feature is similar to the security-proxy's "SP trust SP" one here: https://github.com/georchestra/georchestra/blob/master/security-proxy/src/main/java/org/georchestra/security/ProxyTrustAnotherProxy.java

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/filter/headers/RemoveHeadersGatewayFilterFactory.java:
```diff
@@ -78,7 +78,7 @@ public GatewayFilter apply(RegExConfig regexConfig) {
             final RegExConfig config = regexConfig;// == null ? DEFAULT_SECURITY_HEADERS_CONFIG : regexConfig;
             HttpHeaders incoming = exchange.getRequest().getHeaders();
-            if (config.anyMatches(incoming)) {
+            if (config.anyMatches(incoming) && !incoming.containsKey("sec-mellon-name-id")) {
```
I think the name sounds too specific, we may want something more "generic". 

Also in the previous security-proxy we were relying on the host (name / IP address) from where the request was coming (see https://github.com/georchestra/georchestra/blob/master/security-proxy/src/main/java/org/georchestra/security/ProxyTrustAnotherProxy.java#L62-L85), instead of a header, which sounds safer:
what if the user sends a "sec-mellon-name-id" header in ones http request ? I'd expect he/she will be trusted. It sounds safer to check the request is coming from apache.

pmauduit replied:
this has been fixed in the following commits, but still require some documentation on how to configure the filter (`sec-georchestra-preauthenticated: true` header, and so on...)

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/GatewaySecurityConfiguration.java:
```diff
@@ -61,9 +76,15 @@ public class GatewaySecurityConfiguration {
     @Autowired(required = false)
     ServerLogoutSuccessHandler oidcLogoutSuccessHandler;
 
+    @Autowired(required = false)
+    private AccountDao accountDao;
+
+    @Autowired(required = false)
+    private RoleDao roleDao;
+
     @Bean
     public SecurityWebFilterChain securityWebFilterChain(ServerHttpSecurity http,
-            List<ServerHttpSecurityCustomizer> customizers) throws Exception {
+            List<org.georchestra.gateway.security.ServerHttpSecurityCustomizer> customizers) throws Exception {
```
why fully-qualifying the class ?

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/GatewaySecurityConfiguration.java:
```diff
@@ -103,6 +124,10 @@ private Stream<ServerHttpSecurityCustomizer> sortedCustomizers(List<ServerHttpSe
         return new ResolveGeorchestraUserGlobalFilter(resolver);
     }
 
+    public @Bean ResolveApacheGeorchestraUserFilter resolveApacheGeorchestraUserFilter() {
+        return new ResolveApacheGeorchestraUserFilter();
+    }
```
Apache sounds specific, of course this is what we will actually have with mod_mellon, but we could imagine use this mechanism with whatever we have in front of the gateway (nginx, ...), maybe `ResolveHttpHeadersGeorchestraUserFilter` ?

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/ResolveApacheGeorchestraUserFilter.java:
```diff
@@ -0,0 +1,151 @@
+
+@RequiredArgsConstructor
+@Slf4j(topic = "org.georchestra.gateway.security")
+public class ResolveApacheGeorchestraUserFilter implements GlobalFilter, Ordered {
```
Same remark on Apache being a bit too specific here

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/ResolveApacheGeorchestraUserFilter.java:
```diff
@@ -0,0 +1,151 @@
+    }
+    
+    public @Override Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+        if (exchange.getRequest().getHeaders().containsKey("sec-mellon-name-id")) {
```
See comment above, relying on the host issuing the HTTP request sounds safer, as we can trust where it comes from

pmauduit reviewed (changes requested):
It looks as a good start, thanks @marwanehcine ! some comments though

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/autoconfigure/app/FiltersAutoConfiguration.java:
```diff
@@ -65,4 +69,9 @@ public class FiltersAutoConfiguration {
         return new StripBasePathGatewayFilterFactory();
     }
 
+    @Bean
+    RouterFunction<ServerResponse> staticResourceRouter() {
+        return RouterFunctions.resources("/**", new ClassPathResource("static/"));
+    }
+
```
what is the purpose of introducing this bean ?

marwanehcine replied:
static folder which contains css files is not accessible by default on gateway. We need this bean to make it accessible when redirection is set to gateway by Apache proxy.

pmauduit replied:
so it's more of a bugfix unrelated to the current PR, ok, I was not aware we already had a static folder in the classpath.

marwanehcine replied:
static folder is accessible if no Apache redirection is set, i think this bug is regression when using Apache Proxy.

pmauduit replied:
> static folder is accessible if no Apache redirection is set, i think this bug is regression when using Apache Proxy.


I think this is specific to the private docker composition (apache / traefik config) that I wrote to make the development easier (https://github.com/camptocamp/inrae-renater-docker/blob/master/docker-compose.yaml#L65-L82) ; if we already have a /static route in the gateway and it works, then we don't probably need the bean to be defined, but we could check this when deploying an early modified version of the gateway.

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/filter/headers/RemoveSecurityHeadersGatewayFilterFactory.java:
```diff
@@ -49,12 +50,13 @@ public class RemoveSecurityHeadersGatewayFilterFactory extends AbstractGatewayFi
             DEFAULT_SEC_HEADERS_PATTERN);
 
-    public RemoveSecurityHeadersGatewayFilterFactory() {
+    public RemoveSecurityHeadersGatewayFilterFactory(GatewayConfigProperties configProps) {
         super(Object.class);
+        delegate = new RemoveHeadersGatewayFilterFactory(configProps);
     }
```
I was wondering why you could not `@autowire` the configProps object, maybe it just does not exist as a bean

marwanehcine replied:
No, there is a dependency issue here , autowired did not work cause there is order mismatch, we need to put it into constructor to say wait for this bean to be there and then instantiate RemoveSecurityHeadersGatewayFilterFactory bean.

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/ResolveHttpHeadersGeorchestraUserFilter.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    public @Override Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+        if (exchange.getRequest().getHeaders().containsKey("sec-mellon-name-id")) {
```
we should rename the header to something more generic than "mellon-name-id"

marwanehcine replied:
Ok, it should be update on Apache config file at first

pmauduit replied:
let's say we will have a `sec-georchestra-preauthenticated: true`

pmauduit commented on gateway/src/main/resources/application.yml:
```diff
@@ -23,7 +23,13 @@ spring:
     web-application-type: reactive
   application.name: gateway-service
-  thymeleaf.cache: false
+  thymeleaf:
+    cache: false
+    check-template: true
+    check-template-location: true
+    enabled: true
+    prefix: classpath:/templates/
+    suffix: .html
```
I am surprised that thymeleaf templates are relocated into the classpath, on the MEL project we located the templates/ subdirectory into the georchestra datadir.
Anyway, I think it is also fine this way, as long as we can still customize the configuration so that an external directory is used.

marwanehcine replied:
Here thymeleaf should be configured due to redirection to gateway by Apache proxy, we need to set suffix (which is most important). In gateway templates is under /ressources which is referenced by classpath.

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/ResolveHttpHeadersGeorchestraUserFilter.java:
```diff
@@ -0,0 +1,146 @@
+                roles.add("ROLE_USER");
+            }
+            userOpt.get().setRoles(roles);
+            userOpt.get().setOrganization("INRAE");
```
this should not be hardcoded

pmauduit replied:
but then maybe we'd need to create an organization based on what Renater provides ? Will we have enough info for this ?

pmauduit replied:
Maybe we should put the users into a default organization to start with (relates to https://github.com/georchestra/georchestra/issues/4017)

pmauduit replied:
> Does Apache header contains organization ?

It could, I'd have liked to go further on this point and check the current geOrchestra service provider definition, unfortunately my renater account has been messed up somehow, and I am in the process of fixing it, which required to open a ticket at the Renater organization :(

groldan reviewed (changes requested):
Hi, I’m seeing a bunch of ldap related imports on classes that have nothing to do with ldap. We need to maintain certain level of decoupling. Think of the ldap components as if they were a plugin on a separate project module. The classes in core can’t depend on the plugin, it’s to be the other way around. Moreover, this will break if ldap is disabled (I’m pretty sure we can disable it in the configuration).
Sorry for not being more specific but I’m on mobile and my laptop decided to not boot the UI. But you can easily check, if a class is importing ldap components and it’s not in an ldap package then it’s wrong. You can define or extend existing interfaces to allow for the implementation of the ldap-specific logic on the ldap “plugin” or subsystem though.

marwanehcine commented:
> Hi, I’m seeing a bunch of ldap related imports on classes that have nothing to do with ldap. We need to maintain certain level of decoupling. Think of the ldap components as if they were a plugin on a separate project module. The classes in core can’t depend on the plugin, it’s to be the other way around. Moreover, this will break if ldap is disabled (I’m pretty sure we can disable it in the configuration). Sorry for not being more specific but I’m on mobile and my laptop decided to not boot the UI. But you can easily check, if a class is importing ldap components and it’s not in an ldap package then it’s wrong. You can define or extend existing interfaces to allow for the implementation of the ldap-specific logic on the ldap “plugin” or subsystem though.

Hello @groldan , @pmauduit , @fvanderbiest , I was investigating this issue, code here is same as "GatewaySecurityConfiguration" where we have introduced account creation on Gateway.
For me the question is : should we allow ldap account creation at gateway level ? I thought the response was yes.
But what was missing is to make dependency optional (whether there is ldap sever configured or not)
To solve this issue, I have added "LdapEnabledCondition" which check if there is at least one ldap server with status enabled.
I have also used this condition when creating beans related to LDAP in GatewaySecurityConfiguration class.
Please check this commit : 
https://github.com/georchestra/georchestra-gateway/pull/58/commits/f17a823f0fdf42177a71d5f127a1afea4dd9ee71

Thanks,

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/oauth2/LdapEnabledCondition.java:
```diff
@@ -0,0 +1,25 @@
+public class LdapEnabledCondition implements Condition {
+
+    @Override
+    public boolean matches(ConditionContext context, AnnotatedTypeMetadata metadata) {
+        Resource resource = new FileSystemResource(
+                System.getProperty("georchestra.datadir", "/etc/georchestra") + "/gateway/security.yaml");
+        YamlMapFactoryBean yaml = new YamlMapFactoryBean();
+        yaml.setResources(resource);
+        LinkedHashMap<String, LinkedHashMap<String, Object>> ldap = JsonPath.read(yaml.getObject().get("georchestra"),
+                "$.gateway.security.ldap");
+        return ldap.values().stream().anyMatch(el -> el.containsKey("enabled") && ((Boolean) el.get("enabled")));
+    }
```
I think the `@Value` annotation is meant for this, I find very strange the need to read config files by ourselves

marwanehcine replied:
@value testé, non fonctionnelle. ça retourne null!

marwanehcine replied:
@pmauduit , je pense que c'est le meme probleme, les conditions s'executent avant la lecture de properties

pmauduit replied:
There definitely must be another way than reading config files

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/ResolveHttpHeadersGeorchestraUserFilter.java:
```diff
@@ -0,0 +1,145 @@
+        user.setOrganization(org);
+        List<String> roles = user.getRoles().stream().map(r -> r.contains("ROLE_") ? r : "ROLE_" + r)
+                .collect(Collectors.toList());
+        if (roles.isEmpty()) {
+            roles.add("ROLE_USER");
+        }
```
I am wondering if not connected users should have the ROLE_USER (as the roles collection is likely to be empty if we are not connected).

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/filter/headers/RemoveHeadersGatewayFilterFactory.java:
```diff
@@ -61,10 +70,20 @@
+    @Autowired
+    private Environment environment;
+
+    @Value("${georchestra.gateway.security.createNonExistingUsersInLDAP:}")
+    private String trusted;
```
I don't get the link between the value in the annotation and the variable name

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/security/GatewaySecurityConfiguration.java:
```diff
@@ -103,6 +117,12 @@ private Stream<ServerHttpSecurityCustomizer> sortedCustomizers(List<ServerHttpSe
         return new ResolveGeorchestraUserGlobalFilter(resolver);
     }
 
+    @ConditionalOnExpression("${georchestra.gateway.headerAuthentication:false}")
```
It would make more sense to have the property onto `georchestra.gateway.security.headerAuthentication` in my opinion

pmauduit commented on gateway/src/main/java/org/georchestra/gateway/filter/headers/RemoveHeadersGatewayFilterFactory.java:
```diff
@@ -78,14 +97,37 @@ public GatewayFilter apply(RegExConfig regexConfig) {
-            if (config.anyMatches(incoming)) {
+
+            if (config.anyMatches(incoming) && (!configProps.isHeaderAuthentication()
+                    || !headerAuthenticated(exchange.getRequest().getRemoteAddress().getAddress().toString()))) {
```
Tested on the target infrastructure, performing a remote-debugging: `getRemoteAddress()` returns the actual client issuing the request, which is provided as the http headers `x-forwarded-*` family, which we can't trust.

pmauduit replied:
this should return the internal IP address of the upstream proxy (apache mod_mellon in our case):

```
exchange.getRequest().getNativeRequest().connectionRemoteAddress()
```

pmauduit replied:
... but there is a mismatch between the service ip & the pods one in our k8s infrastructure, so I can't tell how the check can be reliable. Maybe it could make more sense to just drop this feature from the application, and use the k8s network policies instead ?

groldan commented:
# TODO

- [ ] The following tests are not run anymore `OpenIdConnectUserMapperTest`, `OAuth2SecurityAutoConfigurationTest`. Reason being the following statement added at 3a5d74ce and 3a5d74ce: ```assumeTrue(System.getProperty("console.test.openldap.ldapurl") != null
                && System.getProperty("console.test.openldap.basedn") != null);```. Which changes the original tests to sort of integration tests but the System properties are never set.

groldan commented:
Closing as superseded by #63
